### PR TITLE
feat(builder): compress go binaries

### PR DESCRIPTION
**Pull request by aledbf**

### CODE REVIEW

**Comment by aledbf** [comment]
closes #176

File: Makefile [hunk 1]
```diff
@@ -34,6 +34,7 @@ all:
 
 bootstrap:
 	${DEV_ENV_CMD} glide install
+	go get github.com/pwaller/goupx
```
Comment (arschles) [inline Makefile]: @aledbf is `goupx` available in a docker image somewhere? We'd like to keep the build tools encapsulated inside Docker images as much as possible.

**Comment by aledbf** [comment]
@arschles no but I can create an image for this task

**Comment by arschles** [comment]
@aledbf that'd be great. if you'd like to PR it, we're happy to put it into the https://github.com/deis/docker-go-dev image

**Comment by aledbf** [comment]
@arschles https://github.com/deis/docker-go-dev/pull/11

**Comment by arschles** [comment]
thanks @aledbf !

File: Makefile [hunk 2]
```diff
@@ -44,6 +45,7 @@ glideup:
 build:
 	${DEV_ENV_PREFIX} -e CGO_ENABLED=0 ${DEV_ENV_IMAGE} go build -a -installsuffix cgo -ldflags ${LDFLAGS} -o ${BINARY_DEST_DIR}/boot boot.go || exit 1
 	@$(call check-static-binary,$(BINARY_DEST_DIR)/boot)
+	goupx ${BINARY_DEST_DIR}/boot
```
Comment (bacongobbler) [inline Makefile]: note that this requires devs to install `goupx`. Maybe should this be added into `go-dev-env` instead?

Comment (aledbf) [inline Makefile]: @bacongobbler yes deis/docker-go-dev#11

Comment (bacongobbler) [inline Makefile]: I'd personally wait for that to be merged to minimize this PR. It won't require users to install `upx` on their machines in order for this to work. :)

File: Makefile (hunk 1 above)
Comment (bacongobbler) [inline Makefile]: See  deis/docker-go-dev#11. We'll probably have to bag and tag another revision of `deis/go-dev` once again, but once that's done we can use that.

**Comment by arschles** [comment]
@aledbf now that deis/docker-go-dev#11 is merged, have you tested this?

**Comment by aledbf** [comment]
@arschles yes

File: Makefile [hunk 3]
```diff
@@ -43,7 +43,8 @@ glideup:
 # the build as a `docker build`.
 build:
 	${DEV_ENV_PREFIX} -e CGO_ENABLED=0 ${DEV_ENV_IMAGE} go build -a -installsuffix cgo -ldflags ${LDFLAGS} -o ${BINARY_DEST_DIR}/boot boot.go || exit 1
-	@$(call check-static-binary,$(BINARY_DEST_DIR)/boot)
+	@$(call check-static-binary,$(BINARY_DEST_DIR)/boot)	
+	${DEV_ENV_PREFIX} ${DEV_ENV_IMAGE} goupx ${BINARY_DEST_DIR}/boot || exit 1
```
Comment (bacongobbler) [inline Makefile]: what's the `|| exit 1` here for? Docker should return non-zero if `goupx` errors inside the container.

**Comment by mboersma** [comment]
This does produce a much smaller Docker image that seems to work fine:
```
$ docker images | grep builder
192.168.64.2:5000/deis/builder git-67c9267  5b8b3b57fe92   15 seconds ago  40.29 MB
deis/builder                   git-4c8ade1  5bd4bff278ba   5 days ago      158.6 MB
quay.io/deisci/builder         v2-beta       d9dd67a38c77  5 weeks ago     168.1 MB
$ kubectl exec -it deis-builder-ivtzp -- ls -alh /usr/bin/boot
-rwxr-xr-x 3 root root 4.0M Feb 17 17:07 /usr/bin/boot
$ kubectl --namespace=deis logs -f deis-builder-ivtzp
[info] Key accepted for user test-81.
[info] Channel type: session
[debug] receiving git repo name: '/test-269375864.git', operation: git-receive-pack, fingerprint: 6f:e3:64:a9:d2:04:c9:e7:a3:9b:1d:44:2f:df:50:7c, user: test-81
[debug] creating repo directory /home/git/test-269375864.git
[info] Creating new directory at /home/git/test-269375864.git
[debug] writing pre-receive hook under /home/git/test-269375864.git
[debug] Writing pre-receive hook to /home/git/test-269375864.git/hooks/pre-receive
[info] git-shell -c git-receive-pack 'test-269375864.git'
[debug] Working Dir: /home/git
...
Waiting for git-receive to run.
Waiting for deploy.
[info] Deploy complete.
$ helm install deis-tests
$ kubectl --namespace=deis logs -f deis-tests
...
Apps
/go/src/github.com/deis/workflow-e2e/tests/apps_test.go:204
  with a deployed app
  /go/src/github.com/deis/workflow-e2e/tests/apps_test.go:203
    can transfer the app to another owner
    /go/src/github.com/deis/workflow-e2e/tests/apps_test.go:202
------------------------------

Ran 31 of 57 Specs in 123.605 seconds
SUCCESS! -- 31 Passed | 0 Failed | 26 Pending | 0 Skipped PASS 
```

**Pull request merged**